src/test/compile-fail/edition-keywords-2018.rs

```diff
@@ -0,0 +1,22 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// compile-flags: -Zedition=2018 -Zunstable-options
+// aux-build:edition-kw-macro-2015.rs
+
+#![feature(raw_identifiers)]
+
+#[macro_use]
+extern crate edition_kw_macro_2015;
+
+pub fn main() {
+    let _ = consumes_async!(async);
+    //~^ ERROR no rules expected the token `async`
```

Comment:
I don't think this is correct behavior, macros' left hand sides don't care about keywordy-ness status of tokens they match. They can equally well match tokens from code snippets written in any Rust editions or even in C++, Java or mix of both. (See https://github.com/rust-lang/rust/issues/49520)
So the "pretend these are raw even if they aren't" shortcut doesn't generally work.

The correct solution requires a bit more effort, but not significantly more, thankfully:
- Tweak all instances of `is_keyword` involving edition-dependent keywords (no such instances right now! hurray)
- Make all instances of `is_reserved_ident` (maybe all 13 of them)  edition-dependent if the result makes difference for edition-dependent keywords.

Reply:
I disagree; we're going to ask macro consumers to rustfix usages of `async` in macro invocations as well, because crates may be using those tokens verbatim (slurping them via `tt` and then pasting them out) as well as using them to match keywords.

I find it would be inconsistent if _some_ macros stopped working in the 2018 epoch when fed `async` but not others.

thoughts, @nikomatsakis ?

Reply:
Also, FWIW for the lint we're going to have to tristate the bool field anyway, as a "specified raw, keyword but not on current epoch, and keyword on current epoch", and we can have the macro matcher consider the last two to be equal. But I'm not certain.

Reply:
This is not only about macros, what bothers me is that the "pretend these are raw" implementation works on fundamentally wrong level. There are no keywords in the token world, that's why neither lexer nor macros weren't concerned about keywords previously. Keywords appear only when we introduce some grammar on those tokens, so the edition-keyword problem need to be solved on grammar/syntax level as well.
This means `is_reserved_ident` locally and *I think* deserialization (or/and maybe expansion?) of macros from other crates. Any cross-crate-cross-edition compatibility for keywords would involve some amount of guessing (including this PR's) but we should be able to guess better if we are doing it at the right level.
I need some time to flesh this out, maybe make a prototype (this shouldn't be a large change), I'm not satisfied with what this PR does at all.

Reply:
Okay, so --- in the changes I haven't made yet (but planned to for the linting step once we figure that out better), there will be three states:

 - actually raw
 - allowed to be an ident (specified as non-raw on an edition where it is not keyword)
 - not allowed to be an ident (specified as non-raw on an edition where it is a keyword)

This will mean that we no longer pretend things are raw, and we can do equality differently in different contexts.


For epoch hygiene to work we basically have to do this at the token level, though. I don't like having keywords pollute the lexer, but I don't see a better way of doing this that handles macros well.

Reply:
We should decide precisely what the semantics should be here, and how to implement it, before taking more steps.

Reply:
>I need some time to flesh this out, maybe make a prototype

ETA: This weekend (or next week at worst).

Reply:
Any updates?

Reply:
Sorry for the delay.
I know how to solve this - use hygiene, literally.
I.e. check the "keywordness" property of an identifier in its hygienic context (`ident.span.ctxt()`). This is a simple principled solution that doesn't interfere with orthogonal things like lexer, macros or raw identifiers.
I still need to write some implementation boilerplate though, I'll do it this weekend.

Reply:
Sounds good. I kinda investigated this a bit and I was somewhat iffy around the interaction with macros, but you know more than I do here :smiley: . Seems like you'd be doing these keywordness checks in the parser (as usual) and doing the lints alongside those checks. That works okayish. We still need to track lint levels in the parser, though.

(Or do you have a solution for how we should be linting this?)

Reply:
We'd also need to lint stuff in macro defs and invocations, which complicates things.

As an edition breakage it's important that we lint everything.

Reply:
Regarding reporting "really everything", I think this is more of a problem with lint infrastructure than edition keywords specifically.

Reply:
Agreed, but it's not clear how to fix it.

The proposal we had was that the [parser can keep track of lint levels for specific lints](https://github.com/rust-lang/rust/issues/49716#issuecomment-379099673) and we just lint in the parser.

We can also stash lints by span as suggested by @oli-obk

Reply:
(though that still requires some way of visiting them)